reposerver/repository/repository.go

```diff
@@ -1253,6 +1362,259 @@ func getRepoCredential(repoCredentials []*v1alpha1.RepoCreds, repoURL string) *v
+	if err != nil {
+		logCtx.Errorf("SecJoin(%s, %s) failed: %v", dstRoot, dstPath, err)
+		return err
+	}
+
+	dstStat, err := os.Lstat(dstFullPath)
+	dstExists := false
+	logCtx = logCtx.WithField("dstFullPath", dstFullPath)
+
+	// check destination to be within bounds:
+	if err == nil {
+		dstExists = true
+		// if the target exists check it
+		if err := argopath.CheckPathOutOfBound(dstRoot, "", dstPath); err != nil {
+			logCtx.Debugf("copyFilesystemSubtree: argopath.CheckPathOutOfBound failed: %v", err)
+			return err
+		}
+	} else if errors.Is(err, os.ErrNotExist) {
+		// if the destination doesn't yet exist, check its parent
+		if err := argopath.CheckPathOutOfBound(dstRoot, "", filepath.Dir(dstPath)); err != nil {
+			logCtx.WithField("filepath.Dir(dstPath)", filepath.Dir(dstPath)).Debugf("copyFilesystemSubtree: argopath.CheckPathOutOfBound failed: %v", err)
+			return err
+		}
+	} else {
+		return fmt.Errorf("Error while checking destination path: %v", err)
+	}
+
+	logCtx.WithFields(log.Fields{
+		"dstStat": dstStat,
+		"dstExists": dstExists,
+	}).Debug("copyFilesystemSubtree: checks passed, starting copy logic")
+
+	// if our source is just a single file
+	if srcStat.Mode().IsRegular() {
+		// in case the source is a regular file, we don't have to recurse
+		// Also, if the dst already exists, it's an error
+
+		dstFile := dstFullPath
+		// if the destination exists, it has to be a directory
+		if dstExists {
+			// if it's already there, it has to be a directory
+			if !dstStat.Mode().IsDir() {
+				return fmt.Errorf("Copy destination exists, and it's not a directory")
+			}
+		} else {
+			dstDir := filepath.Dir(dstFullPath)
+			// destination directory doesn't exist
+			if err := os.MkdirAll(dstDir, 0755); err != nil {
+				logCtx.Debugf("copyFilesystemSubtree: os.MkdirAll failed: %v", err)
+				return err
+			}
+
+			// fetch the source file's name
+			fname := filepath.Base(dstFullPath)
+			if fname == "." || fname == string(os.PathSeparator) {
+				logCtx.Errorf("copyFilesystemSubtree: filepath.Base failed: %v", err)
+				return fmt.Errorf("Unable to extract source file's name")
+			}
+
+			//dstFile = filepath.Join(dstDir, fname)
+			dstFile, err = secpath.SecureJoin(dstDir, fname)
+			if err != nil {
+				logCtx.Errorf("SecJoin(%s, %s) failed: %v", dstDir, fname, err)
+				return err
+			}
+		}
+		// Now copy the file over
+		if err := argoutil.CopyFile(srcFullPath, dstFile, 0644); err != nil {
+			logCtx.Errorf("copyFilesystemSubtree: CopyFile failed: %v", err)
+			return err
+		}
+	} else if srcStat.Mode().IsDir() {
+		// now we are copying a directory
+
+		// Make sure that the destination directory exists
+		if err := os.MkdirAll(dstFullPath, 0755); err != nil {
+			logCtx.Debugf("copyFilesystemSubtree: os.MkdirAll failed: %v", err)
+			return err
+		}
+
+		// Copy the content of the source directory to the destination dir
+		err = filepath.WalkDir(srcFullPath, func(path string, d fs.DirEntry, err error) error {
```

Comment:
WalkDir does not follow symbolic links. Is that what we are intending here?

Reply:
No. Symlinks are taken care of in the `copyFilesystemSubtreeWalker` function. Walkdir simple tells the callback the entry of the directory to process (read: lists the directory recursively and calls the callback on each item one by one). This `copyFilesystemSubtreeWalker` function checks the nature of this entry, and handles it accordingly.

WalkDir is expected not to follow symlinks. If it would, the complete `copyFilesystemSubtreeWalker`'s logic had to be rewritten differently.